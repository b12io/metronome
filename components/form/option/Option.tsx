
import React from "react";
import classnames from "classnames";
import { FileHelper } from "../../lib/helpers.es6";

import { Element } from "react";

export type OptionValue = Object | string | number;
type Preview = {
  thumbnail: string | null | undefined;
  original: string | null | undefined;
};
type Props = {
  active?: boolean;
  actionOption?: boolean;
  collectionOption?: boolean;
  contentTypeOption?: boolean;
  disabled?: boolean;
  folderOption?: boolean;
  icon: Element<any> | null | undefined;
  iconRight?: Element<any> | null | undefined;
  label?: string;
  mappingOption?: boolean;
  onOptionClick: (value: OptionValue | null | undefined) => void;
  preview: Preview | null | undefined;
  value?: OptionValue;
  style?: Object;
  children?: Element<any>;
};

function Option({
  actionOption,
  active,
  collectionOption,
  contentTypeOption,
  disabled,
  folderOption,
  icon,
  iconRight,
  label,
  mappingOption,
  onOptionClick,
  value,
  preview,
  style,
  children
}: Props) {
  const iconLeft = mappingOption || folderOption || actionOption;
  const fileInfo = preview ? FileHelper(preview) : null;
  const optionPreview = <div className={classnames({
    'ds-form-control-select__dropdown-options-preview': true,
    '--is-full-width': !collectionOption,
    '--is-pdf': fileInfo && fileInfo.isPdf
  })}>
      {preview && preview.thumbnail && <img src={preview.thumbnail} alt={label} />}
      {preview && fileInfo && fileInfo.isSvg && <img src={fileInfo.path} alt={label} />}
      {preview && !preview.thumbnail && fileInfo && !fileInfo.isSvg && <div className="ds-form-control-select__dropdown-options-preview-icon"><span>{fileInfo && fileInfo.extension}</span></div>}
    </div>;

  return <div style={style} className={classnames({
    active,
    'ds-form-control-select__dropdown-options-item': true,
    'ds-form-control-select__dropdown-options-item--action': actionOption,
    'ds-form-control-select__dropdown-options-item--mapping': mappingOption,
    'ds-form-control-select__dropdown-options-item--disabled': disabled,
    'ds-form-control-select__dropdown-options-item--collection': collectionOption
  })} onMouseDown={e => {
    (disabled || actionOption) && e.preventDefault();
    onOptionClick(value);
  }}>
      {preview && collectionOption && optionPreview}
      {contentTypeOption && <div className="content-type-option">{icon}{label}</div>}
      {iconLeft && icon}
      {!contentTypeOption && label}
      {contentTypeOption && iconRight}
      {preview && !collectionOption && optionPreview}
      {children}
    </div>;
}

Option.defaultProps = {
  active: false,
  collectionOption: false,
  contentTypeOption: false,
  disabled: false,
  folderOption: false,
  icon: null,
  mappingOption: false,
  onOptionClick: () => {},
  preview: null,
  style: {},
  value: undefined
};

export default Option;