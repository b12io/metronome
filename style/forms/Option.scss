.ds-form-control-select__dropdown-options {
  max-height: 320px;
  overflow-y: auto;

  &-preview {
    $s: &;
    width: 60px;
    height: 50px;
    margin-right: 10px;
    display: flex;
    justify-content: center;
    align-items: center;
    overflow: hidden;

    > img {
      height: 100%;
    }

    &-icon {
      width: 34px;
      height: 44px;
      background: $ds-ui-dark-color;
      border-radius: $ds-border-radius-base;
      position: relative;

      span {
        background: $ds-ui-darkest-color;
        border-radius: $ds-border-radius-base;
        color: $ds-ui-lightest-color;
        font-size: 10px;
        line-height: 14px;
        font-weight: bold;
        text-align: center;
        text-transform: uppercase;
        position: absolute;
        bottom: 5px;
        right: 5px;
        padding: 0 4px;
        width: 34px;
        overflow: hidden;
        min-height: 14px;
      }
    }

    &.--is-pdf {
      #{$s}-icon span {
        background: #e2564b;
      }
    }

    &.--is-full-width {
      height: auto;
      width: 70%;
      margin: 0 auto;

      img {
        width: 100%;
      }
    }
  }

  &-item {
    color: $ds-display-regular-color;
    padding: 10px;
    border-bottom: 1px solid $ds-ui-regular-color;
    transition: $ds-transition;
    cursor: pointer;

    &--collection {
      display: flex;
      flex-direction: row;
      align-items: center;
    }

    &.ds-form-control-select__dropdown-options-item--mapping {
      color: $ds-brand-light-color;
      font-size: $body-font-size;
      font-weight: $ds-font-weight-bold;

      &.ds-form-control-select__dropdown-options-item--disabled {
        color: $ds-ui-border-color;
        cursor: not-allowed;

        &:hover {
          background: inherit;
        }
      }
    }

    &.ds-form-control-select__dropdown-options-item--action {
      color: $ds-brand-regular-color;
    }

    &:first-child {
      border-radius: $ds-border-radius-base $ds-border-radius-base 0 0;
    }

    &:last-child {
      border-bottom: 0;
      border-radius: 0 0 $ds-border-radius-base $ds-border-radius-base;
    }

    &:hover {
      background: $ds-ui-regular-color;
    }
  }
}
