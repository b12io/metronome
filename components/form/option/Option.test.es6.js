/* global expect, jest */

import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'

import Option from './Option.es6.js'

it('should have correct className', () => {
  const component = shallow(<Option label="test label" />)
  expect(
    component.hasClass('ds-form-control-select__dropdown-options-item')
  ).toBeTruthy()
})

it('should add collection className for collection option', () => {
  const component = shallow(<Option collectionOption label="test label" />)
  expect(
    component.hasClass(
      'ds-form-control-select__dropdown-options-item--collection'
    )
  )
})

it('should display its label', () => {
  const component = shallow(<Option label="test label" />)
  expect(component.text().trim()).toEqual('test label')
})

it('should call `onOptionClick`', () => {
  const onOptionClick = jest.fn()
  const component = shallow(
    <Option onOptionClick={onOptionClick} label="test label" />
  )
  component
    .find('.ds-form-control-select__dropdown-options-item')
    .simulate('mousedown')
  expect(onOptionClick.mock.calls.length).toEqual(1)
})

it('renders correctly', () => {
  const tree = renderer.create(<Option label="test label" />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders correctly with preview', () => {
  const tree = renderer
    .create(<Option label="test label" preview="test preview" />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders correctly as collection option with preview', () => {
  const preview = {
    thumbnail: 'test-image.jpg',
    original: 'test-image.jpg'
  }
  const tree = renderer
    .create(<Option collectionOption label="test label" preview={preview} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders correctly as collection option without preview', () => {
  const tree = renderer
    .create(<Option collectionOption label="test label" />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
