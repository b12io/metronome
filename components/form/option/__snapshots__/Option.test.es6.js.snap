// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="ds-form-control-select__dropdown-options-item"
  onMouseDown={[Function]}
  style={Object {}}
>
  test label
</div>
`;

exports[`renders correctly as collection option with preview 1`] = `
<div
  className="ds-form-control-select__dropdown-options-item ds-form-control-select__dropdown-options-item--collection"
  onMouseDown={[Function]}
  style={Object {}}
>
  <div
    className="ds-form-control-select__dropdown-options-preview"
  >
    <img
      alt="test label"
      src="test-image.jpg"
    />
  </div>
  test label
</div>
`;

exports[`renders correctly as collection option without preview 1`] = `
<div
  className="ds-form-control-select__dropdown-options-item ds-form-control-select__dropdown-options-item--collection"
  onMouseDown={[Function]}
  style={Object {}}
>
  test label
</div>
`;

exports[`renders correctly with preview 1`] = `
<div
  className="ds-form-control-select__dropdown-options-item"
  onMouseDown={[Function]}
  style={Object {}}
>
  test label
  <div
    className="ds-form-control-select__dropdown-options-preview --is-full-width"
  >
    <div
      className="ds-form-control-select__dropdown-options-preview-icon"
    >
      <span>
        
      </span>
    </div>
  </div>
</div>
`;
